### ISSUE TITLE
Clean up text types

### ISSUE THREAD

**Issue by dpsutton** [metabase/metabase#22717]
We sometimes conflate the proper database types with what types we use over the wire and for display purposes. A perfect example of this is the `:type/UUID` type.

```clojure
postgres=> (ancestors :type/UUID)
#{:type/Text :type/*}
```

In postgress, the UUID type is just 128 bits but we are used to the string representation. This causes bugs like https://github.com/metabase/metabase/pull/22667 where we have confused this common representation of uuids, strings like `"6b02cc81-d050-4b6d-adee-ab946b3b387b"`. We of course send this as a string to the front end, but this is a convenient representation. And it is an error for the frontend to suggest string operations on it because it serializes so nicely.

This manifests in other ways as well. 

```clojure
postgres=> (descendants :type/Text)
#{:type/URL
  :type/PostgresEnum
  :type/State
  :type/Structured  ;; bug
  :type/Name
  :type/SerializedJSON ;; bug (maybe, even more complicated)
  :type/Email
  :type/City
  :type/Title
  :type/Description
  :type/UUID   ;; discussed
  :type/Country
  :type/XML   ;; bug
  :type/JSON  ;; bug
  :type/AvatarURL
  :type/ImageURL
  :type/Comment
  :type/ZipCode}
```

Some of these are properly semantic types that still seem to be in our type hierarchy.

The others listed are structured data (xml, json, UUID) that happen to have a nice serialization form but that is completely separate from their nature. And because of this type issue we'll have casting issues. The linked PR has the same bug with xml, and json.

While here I want to highlight an issue with our handling of json. We treat all JSON as strings.

```sql

create table jason (id serial primary key not null, payload_json json, payload_text text);
insert into jason (payload_json, payload_text) values ('{"a": 1}', '{"a": 1}'), ('{"a": 2}', '{"a": 2}');

-- select our type information after syncing
metabase=# select f.database_type, f.base_type, f.effective_type, f.semantic_type from metabase_field f left join metabase_table t on f.table_id = t.id where t.name = 'jason';
 database_type |    base_type    | effective_type  |    semantic_type
---------------+-----------------+-----------------+---------------------
 json          | type/Structured | type/Structured | type/SerializedJSON
 text          | type/Text       | type/Text       | type/SerializedJSON
 serial        | type/Integer    | type/Integer    | type/PK
 integer       | type/Integer    | type/Integer    | type/Category
(4 rows)
```

I'm not sure if the application tracks which are actually string backed json columns (like our application db) and which are actual json objects. I think this issue will get more serious as we go on.

**Comment by thebiglabasky** [comment]
What would be the suggestion here? I'm trying to figure out if we're talking about some additional concept to express that a type can be "represented as" another basic type (like String for UUIDs for example), or if we should be able to work this out without introducing anything specific?

**Comment by thebiglabasky** [comment]
@ranquild I could use your eyes on this one too as I think it aligns well with the work we're planning on type coercions (at least that's functionally quite related)

**Comment by ranquild** [comment]
@thebiglabasky the problem is that we have a single type hierarchy for both "real" and "semantic types". So `type/Number` is both a real and a semantic type. This complicates things sometimes. Ideally these 2 hierarchies should be separate.

**Comment by thebiglabasky** [comment]
Makes sense at a high level. I’m wondering how we should slice this up into steps if that makes sense to tackle as part of this milestone around coercion strategies?

**Comment by thebiglabasky** [comment]
For now I'd go with two suggestions (added as sub-issues):

1. Separate the types hierarchy into two (data types, semantic types) [SEM-143](https://linear.app/metabase/issue/SEM-143/separate-data-type-hierarchy-from-the-semantic-type-one)
2. Remove filtering options for data types we don't support <- that one I'm much less certain about and is worth discussing further [SEM-144](https://linear.app/metabase/issue/SEM-144/remove-filtering-options-on-fields-that-happen-not-to-be-natively)

**Comment by ranquild** [comment]
@thebiglabasky for 2), it's always possible to filter IS NULL / NOT NULL, and we have a "default" filter widget that allows that, in the query builder at least.

**Comment by thebiglabasky** [comment]
OK so I'll set it to "limiting filter options to the default one" for these types 👍
